caikit/runtime/http_server.py

```diff
@@ -0,0 +1,460 @@
+
+
+class RuntimeHTTPServer(RuntimeServerBase):
+    """An implementation of a FastAPI server that serves caikit runtimes"""
+
+    ###############
+    ## Interface ##
+    ###############
+
+    def __init__(self, tls_config_override: Optional[aconfig.Config] = None):
+        super().__init__(get_config().runtime.http.port, tls_config_override)
+        # if not set in config, timeout is None and unvicorn accepts None or number of seconds
+        unvicorn_timeout_graceful_shutdown = get_config().runtime.http.timeout
+
+        self.app = FastAPI()
+
+        # Set up the central predict servicer
+        inference_service = ServicePackageFactory().get_service_package(
+            ServicePackageFactory.ServiceType.INFERENCE,
+        )
+        self.global_predict_servicer = GlobalPredictServicer(inference_service)
+
+        # Set up the central train servicer
+        # TODO: uncomment later on
+        # train_service = ServicePackageFactory().get_service_package(
+        #     ServicePackageFactory.ServiceType.TRAINING,
+        # )
+        # self.global_train_servicer = GlobalTrainServicer(train_service)
+        self.package_name = inference_service.descriptor.full_name.rsplit(".", 1)[0]
+
+        # Bind all routes to the server
+        self._bind_routes(inference_service)
+        # self._bind_routes(train_service)
+
+        # Parse TLS configuration
+        tls_kwargs = {}
+        if (
+            self.tls_config
+            and self.tls_config.server.key
+            and self.tls_config.server.cert
+        ):
+            log.info("<RUN10001905I>", "Running with TLS")
+            tls_kwargs["ssl_keyfile"] = self.tls_config.server.key
+            tls_kwargs["ssl_certfile"] = self.tls_config.server.cert
+            if self.tls_config.client.cert:
+                log.info("<RUN10001809I>", "Running with mutual TLS")
+                tls_kwargs["ssl_ca_certs"] = self.tls_config.client.cert
+                tls_kwargs["ssl_cert_reqs"] = ssl.CERT_REQUIRED
+
+        # Start the server and run forever
+        config = uvicorn.Config(
+            self.app,
+            host="0.0.0.0",
+            port=self.port,
+            log_level=None,
+            log_config=None,
+            timeout_graceful_shutdown=unvicorn_timeout_graceful_shutdown,
+            **tls_kwargs,
+        )
+        self.server = uvicorn.Server(config=config)
+        self.uvicorn_server_thread = threading.Thread(target=self.server.run)
+
+    def __del__(self):
+        if get_config().runtime.metering.enabled:
+            self._stop_metering()
+
+    # Override context manager impl
+    def __enter__(self):
+        self.start()
+        return self
+
+    # TODO: what should actually happen here?
+    def __exit__(self, type_, value, traceback):
+        pass
+
+    def start(self):
+        """Start the server (blocking)"""
+        self.server.run()
+
+    @contextmanager
+    def run_in_thread(self):
```

Comment:
Nice! I think this should actually be merged into `start` and `stop` as follows:

* Add `blocking: bool = True` as an argument to `start` to match how the `grpc` server does it
* If `blocking` is `False`, launch with a thread like you're doing here (also move the creation of the thread inside the condition so that we don't create an unnecessary thread when running with blocking)
* In `stop`, do the `self.server.should_exit = True` and if `self.uvicorn_server_thread is not None`, also `join` it

With this, you can remove the implementations of `__enter__` and `__exit__` here and rely on the ones in the base class!